osu.Game/Screens/Menu/OnlineMenuBanner.cs

```diff
@@ -0,0 +1,247 @@
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            Current.BindValueChanged(loadNewImages, true);
+            checkForUpdates();
+        }
+
+        private void checkForUpdates()
+        {
+            if (!FetchOnlineContent)
+                return;
+
+            var request = new GetMenuContentRequest();
+            Task.Run(() => request.Perform())
+                .ContinueWith(r =>
+                {
+                    if (r.IsCompletedSuccessfully)
+                        Schedule(() => Current.Value = request.ResponseObject);
+
+                    // if the request failed, "observe" the exception.
+                    // it isn't very important why this failed, as it's only for display.
+                    // the inner error will be logged by framework mechanisms anyway.
+                    if (r.IsFaulted)
+                        _ = r.Exception;
+
+                    Scheduler.AddDelayed(checkForUpdates, TimeSpan.FromMinutes(5).TotalMilliseconds);
+                });
+        }
+
+        /// <summary>
+        /// Takes <see cref="Current"/> and materialises and displays drawables for all valid images to be displayed.
+        /// </summary>
+        /// <param name="images"></param>
+        private void loadNewImages(ValueChangedEvent<APIMenuContent> images)
+        {
+            nextDisplay?.Cancel();
+
+            cancellationTokenSource?.Cancel();
+            cancellationTokenSource = null;
+
+            // A better fade out would be nice, but the menu content changes *very* rarely
+            // so let's keep things simple for now.
+            content.Clear(true);
+
+            LoadComponentsAsync(images.NewValue.Images.Select(i => new MenuImage(i)), loaded =>
+            {
+                if (!images.NewValue.Equals(Current.Value))
+                    return;
+
+                // start hidden
+                foreach (var image in loaded)
+                    image.Hide();
+
+                content.AddRange(loaded);
+
+                displayIndex = -1;
+                showNext();
+            }, (cancellationTokenSource ??= new CancellationTokenSource()).Token);
+        }
+
+        private void showNext()
+        {
+            nextDisplay?.Cancel();
+
+            // If the user is hovering a banner, don't rotate yet.
+            bool anyHovered = content.Any(i => i.IsHovered);
+
+            if (!anyHovered)
+            {
+                int previousIndex = displayIndex;
+
+                // To handle expiration simply, arrange all images in best-next order.
+                // Fade in the first valid one, then handle fading out the last if required.
+                var currentRotation = content
+                                      .Skip(Math.Max(0, previousIndex) + 1)
+                                      .Concat(content.Take(Math.Max(0, previousIndex) + 1));
```

Comment:
Nit but this is wrong at the very start no? Taking `Math.Max(0, previousIndex)` _before_ incrementing makes it so that the rotation always starts from the *second* image in the json rather than the first (`Math.Max(0, -1) + 1`).

This is apparent in the test scene btw.